src/daemon/device_bragi.c

```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
     // Since we use usbrecv for everything, the devices tell us when they are ready to handle another packet.
     return;
 }
+
+static void bragi_generate_encryption_key(uchar* encryption_key) {
+  ckb_generate_random(16, encryption_key);
+  for(int i = 0; i < 16; i++) {
+    encryption_key[i] %= 128;
+  }
+}
+
+static void bragi_set_device_keys(usbdevice* kb, unsigned char* pairing_id, unsigned char* encryption_key) {
+
+  uchar pkt[BRAGI_JUMBO_SIZE] = {0};
+
+  // Write the PAIRING_ID to the device first
+  bragi_open_handle(kb, BRAGI_GENERIC_HANDLE, BRAGI_RES_PAIRINGID);
+  memcpy(pkt + 7, pairing_id, sizeof(uchar) * 8);
+  bragi_write_to_handle(kb, pkt, BRAGI_GENERIC_HANDLE, sizeof(pkt), sizeof(uchar) * 8);
+  bragi_close_handle(kb, BRAGI_GENERIC_HANDLE);
+
+  // The ENCRYPTION_KEY must be writting directly after the PAIRING_ID
+  bragi_open_handle(kb, BRAGI_GENERIC_HANDLE, BRAGI_RES_ENCRYPTIONKEY);
+  memcpy(pkt + 7, encryption_key, sizeof(uchar) * 16);
+  bragi_write_to_handle(kb, pkt, BRAGI_GENERIC_HANDLE, sizeof(pkt), sizeof(uchar) * 16);
+  bragi_close_handle(kb, BRAGI_GENERIC_HANDLE);
+}
+
+void cmd_pair_bragi(usbdevice* kb, usbmode* mode, int dummy1, int dummy2, const char* to) {
+  (void)mode;
+  (void)dummy1;
+  (void)dummy2;
+
+  uchar encryption_key[16];
+  bragi_generate_encryption_key(encryption_key);
+  ckb_info("Generated new encryption key");
+
+  uchar pairing_id[8];
+  ckb_generate_random(8, pairing_id);
+  ckb_info("Generated new pairing id");
+
+  // Set the dongle device pairing_id and encryption_key
+  bragi_set_device_keys(kb, pairing_id, encryption_key);    
+
+  char* devices = strdup(to);
+  char *token_ptr;
+  char* device_name = strtok_r(devices, ",", &token_ptr);
+  while(device_name != NULL) {
+    
+    int ckb_id = 0;
+    if(!sscanf(device_name, "ckb%d", &ckb_id)) {
+      ckb_err("Failed to parse device name: %s", device_name);
+      return;
+    }
+
+    usbdevice* device = keyboard + ckb_id; 
```

Comment:
Please lock the dmutex before accessing any part of `device`.

Reply:
I'm not very familiar with "dmutex", can you point me out some documentation / tips about it?